pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go

```diff
@@ -299,6 +303,47 @@ func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error
 	return nil
 }
 
+func (e *etcdrenew) validateEtcdCertsRenewed(ctx context.Context) error {
```

Comment:
I see that there is a commit for adding a new test to pkg/frontend/admin_openshiftcluster_etcdcertificaterenew_test.go
but I think it would be a better idea to add a unit test for this function.

The problem is that we are testing the functionality of the business logic (certs validation) creating the whole infra structs and mocks when it could be simplified a bit.

From a testing perspective, in validateEtcdCertsRenewed(...) you could first take all the secrets, store them in an slice, or in a map (from secretName to secret value) and then iterate over the map and do the whole validation. Doing that, we make the function more "pure" avoiding mocking e.k and we can just focus on the inputs and outputs and not side effects. We can also, if we want, get a counter of how many certs were not correctly renewed or even the names of those. The important part is making the function more testable without the mock overhead, in the below example I use the counter variant but we can use whatever we want (like names using a slice of strings for the invalid ones).

something like
```go
//definition of your original function
func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error { 
	e.log.Infoln("validating if etcd certs exists and expiry")

	for _, secretname := range e.secretNames {
		e.log.Infof("validating secret %s", secretname)
		cert, err := e.k.KubeGet(ctx, "Secret", namespaceEtcds, secretname)
		if err != nil {
			return err
		}

		var u unstructured.Unstructured
		var secret corev1.Secret
		if err = json.Unmarshal(cert, &u); err != nil {
			return err
		}
		err = kruntime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &secret)
		if err != nil {
			return err
		}
		_, certData, err := utilpem.Parse(secret.Data[corev1.TLSCertKey])
		if err != nil {
			return err
		}
		if utilcert.IsCertExpired(certData[0]) {
			return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", "secret %s is already expired, quitting.", secretname)
		}
	}
	e.log.Infoln("Etcd certs exits, are not expired")

	return nil
}

// this is the pure function we really want to test without the mocks
func (e *etcdrenew) validateEtcdCertsRenewed(ctx context.Context) error { 
	e.log.Infoln("validating if etcd certs are renewed")

	certs := make(map[string][]byte)
	for _, secretname := range e.secretNames {
		e.log.Infof("getting secret %s", secretname)
		cert, err := e.k.KubeGet(ctx, "Secret", namespaceEtcds, secretname)
		if err != nil {
			return err
		}
		certs[secretname] = cert
	}

	notRenewedCerts, err := getNotRenewedCerts(e.log, certs)
	if err != nil {
		return err
	}

	if notRenewedCerts > 0 {
		errorMsg := fmt.Sprintf("%v certificates were not renewed", notRenewedCerts)
		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", errorMsg)
	}

	e.log.Infoln("etcd certificates are successfully renewed")
	return nil
}
```
You will need "encodeSecret" from ~/git/ARO-RP/pkg/frontend/admin_openshiftcluster_etcdcertificaterenew_test.go to test the previous function.

So the main idea is to try to avoid testing functionality in code that includes mocks and just use the mocks to ensure that given X I get Y, but NOT given X I get Y and the functionality should be Z, that last thing should be tested independently from the mocks if possible. 

I think in that way our codebase would be much healthier as we would need less mocks to test business logic.
🙂